File: go/fugue/internal/conn/conn.go
// Package conn manages a single fugue WebSocket connection.
// It multiplexes gRPC streams over the connection and serialises writes.
package conn

import (
	"context"
	"encoding/binary"
	"fmt"
	"log/slog"
	"sync"
	"sync/atomic"

	"github.com/coder/websocket"
	framev1 "github.com/fugue-rpc/fugue/grpcws/frame/v1"
	"github.com/fugue-rpc/fugue/frame"
	"github.com/fugue-rpc/fugue/internal/stream"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/proto"
)

// frameBufPool recycles the per-frame encode buffer to avoid a heap alloc on
// every WriteFrame call. Safe because ws.Write is synchronous — the buffer is
// not referenced after Write returns.
var frameBufPool = sync.Pool{
	New: func() any {
		b := make([]byte, 0, frame.HeaderSize+256)
		return &b
	},
}

// writeItem is a single encoded frame waiting to be sent by the writer goroutine.
// done is nil for fire-and-forget frames (MSG); non-nil for frames that require
// delivery confirmation before the caller unblocks (HEADER, END).
type writeItem struct {
	buf  *[]byte      // pooled; returned to frameBufPool after ws.Write
	done chan<- error // nil → fire-and-forget; non-nil → synchronous write
}

// Conn manages one WebSocket connection and the streams multiplexed over it.
type Conn struct {
	ws          *websocket.Conn
	streams     sync.Map // map[uint32]*stream.Stream
	highestID   uint32   // only written by Serve's single read loop, no lock needed
	writeQueue  chan writeItem
	connClosed  chan struct{} // closed when Serve is shutting down
	log         *slog.Logger

	// OnStream is called in a new goroutine for every incoming BEGIN frame.
	// If nil, incoming streams are accepted but no handler runs.
	OnStream func(s *stream.Stream)

	// Performance limits — set before calling Serve.
	RecvBufSize int // per-stream inbound buffer depth (0 → stream.DefaultRecvBufSize)
	MaxStreams  int // max concurrent streams (0 → unlimited)

	activeStreams         atomic.Int32
	estimatedPayloadSize atomic.Uint32 // EMA of recent MSG payload sizes for buffer pre-growth
}

func New(ws *websocket.Conn, log *slog.Logger) *Conn {
	if log == nil {
		log = slog.Default()
	}
	// Raise the WebSocket read limit to match our MaxPayloadSize (4 MiB payload
	// + 9-byte header). The coder/websocket default of 32 KiB would otherwise
	// reject legitimate large messages before our own size check fires.
	ws.SetReadLimit(int64(frame.MaxPayloadSize) + int64(frame.HeaderSize))
	return &Conn{ws: ws, log: log}
}

// Serve runs the read loop until the connection closes or ctx is cancelled.
// It starts a writer goroutine before entering the read loop and waits for
// it to drain before cancelling streams.
func (c *Conn) Serve(ctx context.Context) error {
	// Size the queue to 2×MaxStreams (minimum 128) so concurrent sends rarely
	// stall waiting for the writer goroutine to catch up.
	queueSize := 128
	if c.MaxStreams > 0 && 2*c.MaxStreams > queueSize {
		queueSize = 2 * c.MaxStreams
	}
	c.writeQueue = make(chan writeItem, queueSize)
	c.connClosed = make(chan struct{})

	writerDone := make(chan struct{})
	go func() {
		defer close(writerDone)
		// coalesceBuf accumulates multiple frames for a single ws.Write call.
		// Pooled separately; not returned to pool if it grew beyond 256 KiB.
		coalesceBuf := make([]byte, 0, 64*1024)
		var pending []writeItem

		flush := func() {
			if len(pending) == 0 {
				return
			}
			err := c.ws.Write(context.Background(), websocket.MessageBinary, coalesceBuf)
			for _, item := range pending {
				frameBufPool.Put(item.buf)
				if item.done != nil {
					item.done <- err
				}
			}
			// Reset for next batch; discard oversized backing arrays.
			if cap(coalesceBuf) > 256*1024 {
				coalesceBuf = make([]byte, 0, 64*1024)
			} else {
				coalesceBuf = coalesceBuf[:0]
			}
			pending = pending[:0]
		}

		for item := range c.writeQueue {
			coalesceBuf = append(coalesceBuf, *item.buf...)
			pending = append(pending, item)
			// Non-blocking drain: batch additional queued frames up to budget.
		drain:
			for len(coalesceBuf) < 64*1024 && len(pending) < 32 {
				select {
				case next, ok := <-c.writeQueue:
					if !ok {
						break drain
					}
					coalesceBuf = append(coalesceBuf, *next.buf...)
					pending = append(pending, next)
				default:
					break drain
				}
			}
			flush()
		}
	}()

	defer func() {
		// Signal WriteFrame callers that the connection is gone, then drain the
		// queue before cancelling streams so any buffered END frames can be flushed.
		close(c.connClosed)
		close(c.writeQueue)
		<-writerDone
		c.cancelAllStreams()
	}()

	for {
		_, msg, err := c.ws.Read(ctx)
		if err != nil {
			return err
		}
		frames, err := frame.DecodeAll(msg)
		if err != nil {
			_ = c.ws.Close(websocket.StatusProtocolError, "bad frame")
			return fmt.Errorf("conn: frame decode: %w", err)
		}
		for _, f := range frames {
			if err := c.dispatch(ctx, f); err != nil {
				return err
			}
		}
	}
}

func (c *Conn) dispatch(ctx context.Context, f frame.Frame) error {
	switch f.Type {
	case frame.TypeBEGIN:
		return c.handleBEGIN(ctx, f)
	case frame.TypeMSG:
		return c.handleMSG(f)
	case frame.TypeEND:
		return c.handleEND(f)
	case frame.TypeRESET:
		return c.handleRESET(f)
	default:
		_ = c.ws.Close(websocket.StatusProtocolError,
			fmt.Sprintf("unknown frame type 0x%02x", f.Type))
		return fmt.Errorf("conn: unknown frame type 0x%02x", f.Type)
	}
}

func (c *Conn) handleBEGIN(ctx context.Context, f frame.Frame) error {
	id := f.StreamID
	if id == 0 {
		_ = c.ws.Close(websocket.StatusProtocolError, "stream_id 0 is reserved")
		return fmt.Errorf("conn: stream_id 0 is reserved")
	}
	// highestID is ONLY read and written inside handleBEGIN, which is always
	// called from the single Serve read goroutine. No other goroutine touches
	// it, so no synchronization is needed.
	if id <= c.highestID {
		_ = c.ws.Close(websocket.StatusProtocolError, "non-monotonic stream_id")
		return fmt.Errorf("conn: stream_id %d not monotonically increasing (highest: %d)", id, c.highestID)
	}
	c.highestID = id

	// Decode BeginPayload before taking any resource slots.
	var bp framev1.BeginPayload
	if len(f.Payload) > 0 {
		if err := proto.Unmarshal(f.Payload, &bp); err != nil {
			_ = c.ws.Close(websocket.StatusProtocolError, "bad BeginPayload")
			return fmt.Errorf("conn: bad BeginPayload on stream %d: %w", id, err)
		}
	}

	// Check concurrent stream limit before allocating any resources.
	if c.MaxStreams > 0 && int(c.activeStreams.Add(1)) > c.MaxStreams {
		c.activeStreams.Add(-1)
		endPayload, _ := proto.Marshal(&framev1.EndPayload{
			StatusCode:    uint32(codes.ResourceExhausted),
			StatusMessage: "too many concurrent streams",
		})
		// Fire-and-forget: do not block the read loop waiting for delivery.
		c.enqueue(c.encodeFrame(frame.TypeEND, id, endPayload), nil)
		return nil
	}

	if len(bp.Metadata) == 0 {
		// Fast path: avoid allocating an empty map; gRPC treats nil MD identically.
		sCtx, cancel := context.WithCancel(metadata.NewIncomingContext(ctx, nil))
		s := stream.New(sCtx, cancel, id, bp.Method, c.RecvBufSize, c)
		c.streams.Store(id, s)
		c.startHandler(id, s)
		return nil
	}

	md := make(metadata.MD, len(bp.Metadata))
	for k, v := range bp.Metadata {
		md[k] = []string{v}
	}
	sCtx, cancel := context.WithCancel(metadata.NewIncomingContext(ctx, md))
	s := stream.New(sCtx, cancel, id, bp.Method, c.RecvBufSize, c)
	c.streams.Store(id, s)
	c.startHandler(id, s)
	return nil
}

func (c *Conn) startHandler(id uint32, s *stream.Stream) {
	go func() {
		defer func() {
			c.streams.Delete(id)
			if c.MaxStreams > 0 {
				c.activeStreams.Add(-1)
			}
		}()
		if c.OnStream != nil {
			c.OnStream(s)
		}
	}()
}

// handleMSG silently drops frames for unknown/closed streams. A MSG arriving
// after the handler returned is an in-flight race, not a protocol error.
// If the stream's inbound buffer is full, the stream is terminated with
// RESOURCE_EXHAUSTED to preserve stream independence.
func (c *Conn) handleMSG(f frame.Frame) error {
	s, ok := c.loadStream(f.StreamID)
	if !ok {
		return nil
	}
	if !s.Deliver(f.Payload) {
		_ = s.SendEnd(status.Error(codes.ResourceExhausted, "inbound message buffer full"))
		s.CloseRecv()
		s.Cancel()
		c.streams.Delete(f.StreamID)
	}
	return nil
}

// handleEND silently drops frames for unknown/closed streams for the same
// reason as handleMSG.
func (c *Conn) handleEND(f frame.Frame) error {
	s, ok := c.loadStream(f.StreamID)
	if !ok {
		return nil
	}
	s.CloseRecv()
	return nil
}

func (c *Conn) handleRESET(f frame.Frame) error {
	s, ok := c.loadStream(f.StreamID)
	if !ok {
		// RESET for unknown/closed stream is silently dropped per spec §4.4.
		return nil
	}
	s.CloseRecv()
	s.Cancel()
	c.streams.Delete(f.StreamID)
	return nil
}

func (c *Conn) loadStream(id uint32) (*stream.Stream, bool) {
	v, ok := c.streams.Load(id)
	if !ok {
		return nil, false
	}
	return v.(*stream.Stream), true
}

// WriteFrame implements stream.Sink.
//
// For TypeMSG frames the call is fire-and-forget: the frame is enqueued and
// WriteFrame returns immediately. If the connection has already closed, the
// frame is silently dropped (consistent with gRPC streaming semantics — there
// is no per-message delivery guarantee).
//
// For TypeHEADER and TypeEND frames WriteFrame blocks until the frame has been
// written to the wire (or the connection closes), preserving the invariant that
// SendHeader and SendEnd do not return before the remote peer can see the frame.
func (c *Conn) WriteFrame(typ uint8, streamID uint32, payload []byte) error {
	if uint32(len(payload)) > frame.MaxPayloadSize {
		return fmt.Errorf("frame: payload %d bytes exceeds MAX_PAYLOAD_SIZE", len(payload))
	}
	buf := c.encodeFrame(typ, streamID, payload)

	if typ == frame.TypeMSG {
		c.enqueue(buf, nil) // fire-and-forget; errors are silently dropped
		return nil
	}

	// HEADER and END: synchronous — block until the writer goroutine delivers.
	done := make(chan error, 1)
	if err := c.enqueue(buf, done); err != nil {
		return err
	}
	return <-done
}

// WriteMsgProto is the zero-copy fast path for MSG frames. It marshals msg
// directly into a pooled frame buffer, eliminating one heap allocation and one
// memcopy compared to the WriteFrame(TypeMSG) path.
//
// Recognised by stream.SendMsg via a type assertion (stream.protoMsgWriter).
func (c *Conn) WriteMsgProto(streamID uint32, msg proto.Message) error {
	bp := frameBufPool.Get().(*[]byte)
	b := (*bp)[:0]
	// Reserve the 9-byte header placeholder; fill it after marshalling so we
	// know the exact payload length.
	b = append(b, 0, 0, 0, 0, 0, 0, 0, 0, 0)
	// Pre-grow to estimated payload size to avoid reallocation on MarshalAppend.
	est := int(c.estimatedPayloadSize.Load())
	if est == 0 {
		est = 256
	}
	if cap(b) < frame.HeaderSize+est {
		grown := make([]byte, frame.HeaderSize, frame.HeaderSize+est)
		copy(grown, b[:frame.HeaderSize])
		b = grown
	}
	// Marshal proto bytes directly after the header — zero intermediate alloc
	// when the buffer has enough capacity.
	var err error
	b, err = proto.MarshalOptions{}.MarshalAppend(b, msg)
	if err != nil {
		frameBufPool.Put(bp)
		return err
	}
	*bp = b

	payloadLen := uint32(len(b) - frame.HeaderSize)
	// Backfill the 9-byte header now that payload length is known.
	b[0] = frame.TypeMSG
	binary.BigEndian.PutUint32(b[1:5], streamID)
	binary.BigEndian.PutUint32(b[5:9], payloadLen)

	// Update EMA: newEst ≈ (7×old + actual) / 8; clamped to [256, 4096].
	if payloadLen > 0 {
		for {
			old := c.estimatedPayloadSize.Load()
			newEst := (old*7 + payloadLen) / 8
			if newEst < 256 {
				newEst = 256
			} else if newEst > 4096 {
				newEst = 4096
			}
			if c.estimatedPayloadSize.CompareAndSwap(old, newEst) {
				break
			}
		}
	}

	c.enqueue(bp, nil)
	return nil
}

// encodeFrame allocates a pooled buffer and encodes the 9-byte header + payload.
func (c *Conn) encodeFrame(typ uint8, streamID uint32, payload []byte) *[]byte {
	bp := frameBufPool.Get().(*[]byte)
	b := (*bp)[:0]
	b = append(b, typ)
	b = binary.BigEndian.AppendUint32(b, streamID)
	b = binary.BigEndian.AppendUint32(b, uint32(len(payload)))
	b = append(b, payload...)
	*bp = b
	return bp
}

// enqueue sends buf to the write queue. If the connection is closing, the
// buffer is returned to the pool, done (if non-nil) receives an error, and
// the error is returned to the caller. For fire-and-forget callers (done==nil)
// the return value can be ignored.
//
// The recover() guard handles the rare race where the select picks the
// writeQueue send case in the same instant that the queue is closed by Serve's
// deferred cleanup — in that case Go panics on send-to-closed-channel.
func (c *Conn) enqueue(buf *[]byte, done chan<- error) (retErr error) {
	defer func() {
		if r := recover(); r != nil {
			frameBufPool.Put(buf)
			err := fmt.Errorf("fugue: connection closed")
			if done != nil {
				done <- err
			}
			retErr = err
		}
	}()
	select {
	case c.writeQueue <- writeItem{buf: buf, done: done}:
		return nil
	case <-c.connClosed:
		frameBufPool.Put(buf)
		err := fmt.Errorf("fugue: connection closed")
		if done != nil {
			done <- err
		}
		return err
	}
}

func (c *Conn) cancelAllStreams() {
	c.streams.Range(func(_, value any) bool {
		s := value.(*stream.Stream)
		s.CloseRecv() // safe: guarded by sync.Once inside Stream
		s.Cancel()
		return true
	})
}

File: packages/node-server/src/conn.ts
// Connection handler for one fugue WebSocket connection.
// Multiplexes gRPC streams per the wire protocol (docs/wire-format.md).

import type { WebSocket, RawData } from "ws";
import { create, fromBinary, toBinary } from "@bufbuild/protobuf";
import { FrameType, encodeFrame, decodeAll } from "./frame.js";
import {
  BeginPayloadSchema, EndPayloadSchema, HeaderPayloadSchema,
} from "./proto.js";
import type { MethodEntry } from "./service.js";
import type {
  UnaryHandler, ServerStreamHandler, ClientStreamHandler, BidiHandler,
  UnaryServerCall, ServerStreamCall, ClientStreamCall, BidiCall,
} from "./service.js";

// gRPC status codes used internally.
const GRPC_OK                 = 0;
const GRPC_UNIMPLEMENTED      = 12;
const GRPC_INTERNAL           = 13;
const GRPC_RESOURCE_EXHAUSTED = 8;

/**
 * Called before every handler dispatch. Throw a gRPC status error to reject
 * the call (e.g. `throw Object.assign(new Error("UNAUTHENTICATED"), { code: 16 })`).
 * Runs before the method lookup, so it fires even for unimplemented methods.
 */
export type CallInterceptor = (
  ctx: { method: string; metadata: Readonly<Record<string, string>> }
) => void | Promise<void>;

export interface ConnOptions {
  /** Per-stream inbound message buffer depth. When full the stream is RESET. Default: 256. */
  recvBufSize?: number;
  /** Max concurrent streams per connection. 0 = unlimited (default). */
  maxStreams?: number;
  /** Pre-call hook for auth, logging, or metrics. Throw to reject a call. */
  interceptor?: CallInterceptor;
}

// Bounded async-push queue that backs the recv side of each stream.
// Single-consumer: only one [Symbol.asyncIterator] call per instance.
class RecvQueue {
  private readonly _buf: Buffer[] = [];
  private _waiter: (() => void) | null = null;
  private _closed = false;

  constructor(private readonly _capacity: number) {}

  // Returns true if accepted, false if buffer is full (caller should RESET stream).
  push(item: Buffer): boolean {
    if (this._closed) return true;
    if (this._buf.length >= this._capacity) return false;
    this._buf.push(item);
    const w = this._waiter;
    this._waiter = null;
    w?.();
    return true;
  }

  close(): void {
    if (this._closed) return;
    this._closed = true;
    const w = this._waiter;
    this._waiter = null;
    w?.();
  }

  [Symbol.asyncIterator](): AsyncIterator<Buffer> {
    return this._iterate();
  }

  private async *_iterate(): AsyncGenerator<Buffer> {
    while (true) {
      if (this._buf.length > 0) {
        yield this._buf.shift()!;
      } else if (this._closed) {
        return;
      } else {
        await new Promise<void>(r => { this._waiter = r; });
      }
    }
  }
}

interface StreamState {
  recvQueue: RecvQueue;
  headerSent: boolean;
  trailers: Record<string, string>;
  aborted: boolean;
}

export class FugueConn {
  private readonly _streams = new Map<number, StreamState>();
  private _highestId = 0;
  private _activeStreams = 0;
  private readonly _recvBufSize: number;
  private readonly _maxStreams: number;
  private readonly _interceptor: CallInterceptor | undefined;
  private _pendingFrames: Buffer[] = [];
  private _flushScheduled = false;
  private _connClosed = false;

  constructor(
    private readonly _ws: WebSocket,
    private readonly _lookup: (path: string) => MethodEntry | undefined,
    options?: ConnOptions,
  ) {
    this._recvBufSize = options?.recvBufSize ?? 256;
    this._maxStreams = options?.maxStreams ?? 0;
    this._interceptor = options?.interceptor;
  }

  serve(): void {
    this._ws.on("message", (data: RawData, isBinary: boolean) => {
      if (!isBinary) {
        this._ws.close(1002, "text frames are a protocol error");
        return;
      }
      let buf: Buffer;
      if (Buffer.isBuffer(data)) {
        buf = data;
      } else if (Array.isArray(data)) {
        buf = Buffer.concat(data as Buffer[]);
      } else {
        buf = Buffer.from(data as ArrayBuffer);
      }
      this._onMessage(buf);
    });

    this._ws.on("close", () => {
      this._connClosed = true;
      this._pendingFrames = [];
      for (const state of this._streams.values()) {
        state.aborted = true;
        state.recvQueue.close();
      }
    });

    this._ws.on("error", () => {
      // Errors surface as a close event; suppress the unhandled-error crash.
    });
  }

  private _onMessage(data: Buffer): void {
    let frames;
    try {
      frames = decodeAll(data);
    } catch (err) {
      this._ws.close(1002, String(err));
      return;
    }
    for (const frame of frames) {
      if (this._dispatch(frame)) return; // connection closed
    }
  }

  // Returns true if the connection was closed due to a protocol error.
  private _dispatch(frame: { type: number; streamId: number; payload: Buffer }): boolean {
    switch (frame.type) {
      case FrameType.BEGIN: return this._handleBEGIN(frame.streamId, frame.payload);
      case FrameType.MSG:   return this._handleMSG(frame.streamId, frame.payload);
      case FrameType.END:   return this._handleEND(frame.streamId);
      case FrameType.RESET: return this._handleRESET(frame.streamId);
      default:
        this._ws.close(1002, `unknown frame type 0x${frame.type.toString(16).padStart(2, "0")}`);
        return true;
    }
  }

  private _handleBEGIN(streamId: number, payload: Buffer): boolean {
    if (streamId === 0) {
      this._ws.close(1002, "stream_id 0 is reserved");
      return true;
    }
    if (streamId <= this._highestId) {
      this._ws.close(1002, `non-monotonic stream_id: ${streamId} ≤ highest ${this._highestId}`);
      return true;
    }
    this._highestId = streamId;

    let method = "";
    let metadata: Record<string, string> = {};
    if (payload.length > 0) {
      try {
        const bp = fromBinary(BeginPayloadSchema, payload);
        method = bp.method;
        metadata = bp.metadata;
      } catch {
        this._ws.close(1002, "bad BeginPayload");
        return true;
      }
    }

    if (this._maxStreams > 0 && this._activeStreams >= this._maxStreams) {
      this._sendFrame(
        FrameType.END, streamId,
        encodeEndPayload(GRPC_RESOURCE_EXHAUSTED, "too many concurrent streams", {}),
      );
      return false;
    }

    const state: StreamState = {
      recvQueue: new RecvQueue(this._recvBufSize),
      headerSent: false,
      trailers: {},
      aborted: false,
    };
    this._streams.set(streamId, state);
    this._activeStreams++;

    this._dispatchHandler(streamId, method, metadata, state).catch(() => {});
    return false;
  }

  private _handleMSG(streamId: number, payload: Buffer): boolean {
    const state = this._streams.get(streamId);
    if (!state) return false; // in-flight race after stream closed — silently drop
    if (!state.recvQueue.push(payload)) {
      // Buffer full: reset stream, but keep the connection alive (stream independence).
      state.aborted = true;
      state.recvQueue.close();
      this._streams.delete(streamId);
      this._activeStreams--;
      this._sendFrame(
        FrameType.END, streamId,
        encodeEndPayload(GRPC_RESOURCE_EXHAUSTED, "inbound message buffer full", {}),
      );
    }
    return false;
  }

  private _handleEND(streamId: number): boolean {
    const state = this._streams.get(streamId);
    if (!state) return false;
    state.recvQueue.close();
    return false;
  }

  private _handleRESET(streamId: number): boolean {
    const state = this._streams.get(streamId);
    if (!state) return false; // spec §4.4: silently drop RESET for unknown stream
    state.aborted = true;
    state.recvQueue.close();
    this._streams.delete(streamId);
    this._activeStreams--;
    return false;
  }

  private async _dispatchHandler(
    streamId: number,
    method: string,
    metadata: Record<string, string>,
    state: StreamState,
  ): Promise<void> {
    if (this._interceptor) {
      try {
        await this._interceptor({ method, metadata });
      } catch (err) {
        this._cleanup(streamId);
        const { code, message } = extractGrpcStatus(err);
        this._sendFrame(FrameType.END, streamId, encodeEndPayload(code, message, {}));
        return;
      }
    }

    const entry = this._lookup(method);
    if (!entry) {
      this._cleanup(streamId);
      this._sendFrame(
        FrameType.END, streamId,
        encodeEndPayload(GRPC_UNIMPLEMENTED, `unknown method ${method}`, {}),
      );
      return;
    }

    const sendHeader = (headers: Record<string, string>): void => {
      if (state.headerSent || state.aborted) return;
      state.headerSent = true;
      this._sendFrame(
        FrameType.HEADER, streamId,
        toBinary(HeaderPayloadSchema, create(HeaderPayloadSchema, { headers })),
      );
    };

    const autoHeader = (): void => {
      if (!state.headerSent) sendHeader({});
    };

    const setTrailer = (trailers: Record<string, string>): void => {
      Object.assign(state.trailers, trailers);
    };

    const write = (response: unknown): void => {
      if (state.aborted) return;
      autoHeader();
      this._sendFrame(FrameType.MSG, streamId, entry.serialize(response as never));
    };

    const callBase = { metadata, sendHeader, setTrailer };

    let handlerError: unknown = null;
    try {
      switch (entry.kind) {
        case "unary": {
          const request = await this._recvFirst(state.recvQueue, entry.deserialize);
          if (state.aborted) break;
          const call: UnaryServerCall<unknown> = { ...callBase, request };
          const response = await (entry.handler as UnaryHandler<unknown, unknown>)(call);
          if (!state.aborted) {
            autoHeader();
            this._sendFrame(FrameType.MSG, streamId, entry.serialize(response));
          }
          break;
        }
        case "server_stream": {
          const request = await this._recvFirst(state.recvQueue, entry.deserialize);
          if (state.aborted) break;
          const call: ServerStreamCall<unknown, unknown> = { ...callBase, request, write };
          await (entry.handler as ServerStreamHandler<unknown, unknown>)(call);
          break;
        }
        case "client_stream": {
          const iter = makeIter(state.recvQueue, entry.deserialize);
          const call: ClientStreamCall<unknown> = {
            ...callBase,
            [Symbol.asyncIterator]: () => iter,
          };
          const response = await (entry.handler as ClientStreamHandler<unknown, unknown>)(call);
          if (!state.aborted) {
            autoHeader();
            this._sendFrame(FrameType.MSG, streamId, entry.serialize(response));
          }
          break;
        }
        case "bidi_stream": {
          const iter = makeIter(state.recvQueue, entry.deserialize);
          const call: BidiCall<unknown, unknown> = {
            ...callBase,
            write,
            [Symbol.asyncIterator]: () => iter,
          };
          await (entry.handler as BidiHandler<unknown, unknown>)(call);
          break;
        }
      }
    } catch (err) {
      handlerError = err;
    }

    this._cleanup(streamId);

    if (state.aborted) return;

    if (handlerError != null) {
      const { code, message } = extractGrpcStatus(handlerError);
      this._sendFrame(FrameType.END, streamId, encodeEndPayload(code, message, state.trailers));
    } else {
      this._sendFrame(FrameType.END, streamId, encodeEndPayload(GRPC_OK, "", state.trailers));
    }
  }

  private _cleanup(streamId: number): void {
    if (this._streams.delete(streamId)) {
      this._activeStreams--;
    }
  }

  private async _recvFirst(queue: RecvQueue, deserialize: (b: Buffer) => unknown): Promise<unknown> {
    for await (const buf of queue) {
      return deserialize(buf);
    }
    throw Object.assign(new Error("stream closed before request arrived"), { code: GRPC_INTERNAL });
  }

  private _sendFrame(type: number, streamId: number, payload: Uint8Array): void {
    if (this._connClosed) return;
    this._pendingFrames.push(encodeFrame(type, streamId, payload));
    if (!this._flushScheduled) {
      this._flushScheduled = true;
      queueMicrotask(() => { this._flush(); });
    }
  }

  private _flush(): void {
    this._flushScheduled = false;
    const frames = this._pendingFrames;
    if (frames.length === 0 || this._connClosed) return;
    this._pendingFrames = [];
    try {
      this._ws.send(frames.length === 1 ? frames[0] : Buffer.concat(frames));
    } catch {
      // WebSocket already closed — ignore.
    }
  }
}

function makeIter<T>(queue: RecvQueue, deserialize: (b: Buffer) => T): AsyncIterator<T> {
  return (async function* () {
    for await (const buf of queue) {
      yield deserialize(buf);
    }
  })();
}

function encodeEndPayload(code: number, message: string, trailers: Record<string, string>): Uint8Array {
  return toBinary(EndPayloadSchema, create(EndPayloadSchema, {
    statusCode: code,
    statusMessage: message,
    trailers,
  }));
}

function extractGrpcStatus(err: unknown): { code: number; message: string } {
  if (err != null && typeof err === "object") {
    const e = err as Record<string, unknown>;
    if (typeof e["code"] === "number" && typeof e["message"] === "string") {
      return { code: e["code"] as number, message: e["message"] as string };
    }
  }
  return { code: GRPC_INTERNAL, message: String(err) };
}
